### ISSUE TITLE
one hot -> nn.Embedding

### ISSUE THREAD

**Issue by jeff-regier** [YosefLab/scVI-dev#39]
This doesn't seem to run on the GPU. I think we want to use `torch.nn.Embedding` instead of this one hot thing, with `padding_idx=True`.

https://github.com/YosefLab/scVI-dev/blob/b5bd5e18ee72d49262d099771900c62524142975/scvi/scvi.py#L228-L240

**Comment by jeff-regier** [comment]
Actually, it would also work to do
```
        onehot.scatter_(1, batch_ind, 1)
```

I think `batch_index` should be returned from the `dataset.py` as type `np.int64` so it's automatically converted to to a LongTensor by the time it reaches this block of code. 

Also, rather than doing `torch.zeros`, which only runs on the CPU, do
```
onehot = batch_index.new(batch_index.size(0), n_batch)
```
That will run on the GPU is everything else is running on the GPU.

**Comment by Edouard360** [comment]
Now it looks like:

https://github.com/YosefLab/scVI-dev/blob/96483d0b9b000f9c4912a7a9af03c6a0543640ad/scvi/scvi.py#L212-L219]]

I agree batch_index should be returned as a int (though torch.LongTensor rather than np.int64, no ?), so I could probably change:

https://github.com/YosefLab/scVI-dev/blob/e4e69f3e3f788f9c639a79d79445bd15e1d91765/scvi/dataset/dataset.py#L51-L54

to:

    def __getitem__(self, idx):
        # Returns the triplet (X, local_mean, local_var)
        # Shouldn't we call .cuda here ?
        return self.X[idx, :self.nb_genes], self.X[idx, -3:-2], self.X[idx, -2:-1], self.X[idx, -1:].type(torch.LongTensor)

But that might impact performance. (?)

**Comment by jeff-regier** [comment]
In the `get_index` method, if `batch_index` is returned as an int, it gets converted by `DataLoader` to `torch.LongTensor`.

We don't want to store batch_index, etc in the last columns of the X matrix, I think. Much more readible for this class to have separate fields for them. You might add a field called `self.batch_indexes`. Could be a torch.longtensor or np.int64

**Comment by Edouard360** [comment]
Ah ok, I think I see: I can still refer to these different fields in the `__getitem__` function... 
it would look something like below right ?

    def __init__(self, Xs):
        ...
        new_Xs = []
        local_means = []
        local_vars =[]
        batch_indices = []
        for i, X in enumerate(Xs):
            X = torch.Tensor(X)
            log_counts = torch.log(torch.sum(X, dim=1))
            local_mean = torch.mean(log_counts)
            local_var = torch.var(log_counts)
            new_Xs += [X]
            local_means+= [local_mean * torch.ones((X.size()[0], 1))]
            local_vars += [local_var * torch.ones((X.size()[0], 1))]
            batch_indices += [torch.LongTensor(i * np.ones((X.size()[0], 1)))]

        self.local_means = torch.cat(local_means,dim=0)
        self.local_vars = torch.cat(local_vars,dim=0)
        self.batch_indices = torch.cat(batch_indices)
        self.X = torch.cat(new_Xs, dim=0)
        self.total_size = self.X.size(0)

    def __getitem__(self, idx):
        return self.X[idx], self.local_means[idx], self.local_vars[idx], self.batch_indices[idx]

**Comment by jeff-regier** [comment]
Yes, looks good!
